cpp/src/dictionary/remove_keys.cu

```diff
@@ -60,46 +56,61 @@ std::unique_ptr<column> remove_keys_fn(
-
-  // init values to max to identify new nulls
-  rmm::device_uvector<uint32_t> map_indices(keys_view.size(), stream);
+  auto indices_type    = dictionary_column.indices().type();
+  auto max_size        = dictionary_column.size();
+
+  // create/init indices map array
+  auto map_indices =
+    make_fixed_width_column(indices_type, keys_view.size(), mask_state::UNALLOCATED, stream);
+  auto map_itr =
+    cudf::detail::indexalator_factory::make_output_iterator(map_indices->mutable_view());
+  // init to max to identify new nulls
   thrust::fill(execpol->on(stream),
-               map_indices.begin(),
-               map_indices.end(),
-               static_cast<uint32_t>(max_column_size));
-  // copy the non-removed keys ( keys_to_keep_fn(idx)==true );
+               map_itr,
+               map_itr + keys_view.size(),
+               max_size);  // all valid indices are less than this value
+
+  // build keys column and indices map
   std::unique_ptr<column> keys_column = [&] {
-    auto table_keys = cudf::detail::copy_if(
-                        table_view{{keys_view, keys_positions_view}}, keys_to_keep_fn, mr, stream)
-                        ->release();
-    keys_positions_view = table_keys[1]->view();
+    // create keys positions column to identify original key positions after removing they keys
+    auto keys_positions = [&] {
+      auto positions = make_fixed_width_column(
+        indices_type, keys_view.size(), cudf::mask_state::UNALLOCATED, stream);
+      auto itr = cudf::detail::indexalator_factory::make_output_iterator(positions->mutable_view());
+      thrust::sequence(execpol->on(stream), itr, itr + keys_view.size());
+      return std::move(positions);
+    }();
+    // copy the non-removed keys ( keys_to_keep_fn(idx)==true )
+    auto table_keys =
+      cudf::detail::copy_if(
+        table_view{{keys_view, keys_positions->view()}}, keys_to_keep_fn, mr, stream)
+        ->release();
+    auto const filtered_view = table_keys[1]->view();
+    auto filtered_itr = cudf::detail::indexalator_factory::make_input_iterator(filtered_view);
+    auto positions_itr =
+      cudf::detail::indexalator_factory::make_input_iterator(keys_positions->view());
     // build indices mapper
     // Example scatter([0,1,2][0,2,4][max,max,max,max,max]) => [0,max,1,max,2]
     thrust::scatter(execpol->on(stream),
-                    thrust::make_counting_iterator<uint32_t>(0),
-                    thrust::make_counting_iterator<uint32_t>(keys_positions_view.size()),
-                    keys_positions_view.begin<uint32_t>(),
-                    map_indices.begin());
+                    positions_itr,
+                    positions_itr + filtered_view.size(),
+                    filtered_itr,
+                    map_itr);
     return std::move(table_keys.front());
-  }();  // frees up the temporary table_keys objects
+  }();
 
+  // create non-nullable indices view with offset applied -- this is used as a gather map
```

Comment:
Is it safe to ignore the null mask? :thinking:

Reply:
`cudf::detail::gather` requires the indices values to be non-null which makes sense. The null-mask is resolved separately below using the `cudf::detail::valid_if` function.